File: ifoodAndroidTest/app/src/main/java/test/ifood/uellisson/ifoodandroidtest/data/network/TwitterAPI.java
package test.ifood.uellisson.ifoodandroidtest.data.network;

import java.util.List;
import retrofit2.Call;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.GET;
import retrofit2.http.Header;
import retrofit2.http.POST;
import retrofit2.http.Query;
import test.ifood.uellisson.ifoodandroidtest.data.model.TwitterAuthorizationToken;
import test.ifood.uellisson.ifoodandroidtest.data.model.TweetEntity;

public interface TwitterAPI {

    @POST("oauth2/token")
    @FormUrlEncoded
    Call<TwitterAuthorizationToken> getAccessToken(@Header("Authorization") String authorization,
                                                   @Field("grant_type") String grantType);


    @GET("1.1/statuses/user_timeline.json")
    Call<List<TweetEntity>> getTweetsByUsername(
            @Header("Authorization") String authorization,
            @Query("screen_name") String query
    );
}

File: app/src/main/java/com/rlino/ifoodtwitterchallenge/data/twitter/TwitterApi.kt
package com.rlino.ifoodtwitterchallenge.data.twitter

import com.rlino.ifoodtwitterchallenge.BuildConfig
import com.rlino.ifoodtwitterchallenge.model.Tweet
import io.reactivex.Single
import retrofit2.Call
import retrofit2.http.*

interface TwitterApi {

    @GET("${BuildConfig.TWITTER_API_VERSION}/statuses/user_timeline.json")
    fun getTweetsFromUser(@Query("screen_name") from: String,
                          @Query("exclude_replies") excludeReplies: Boolean = true,
                          @Query("include_rts") includeRts: Boolean = false,
                          @Query("count") count: Int = 50
    ): Single<List<Tweet>>

    @POST("oauth2/token")
    @FormUrlEncoded
    fun getToken(
            @Header("Authorization") authorization: String,
            @Field("grant_type") grantType: String = "client_credentials"
    ): Single<TokenResponse>

}